mmdet/models/dense_heads/autoassign_head.py

```diff
@@ -0,0 +1,413 @@
+                prediction box, has shape of (num_points, num_gt).
+            gt_labels (Tensor): The zeros based gt_labels of all gt
+                with shape of (num_gt,).
+            center_prior_weights (Tensor): Float tensor with shape
+                of (num_points, num_gt). Each value represents
+                the center weighting coefficient.
+
+        Returns:
+            tuple:
+            pos_loss (Tensor): The positive loss of all points
+                in the gt_bboxes.
+        """
+        p_loc = torch.exp(-5 * reg_loss)
+        p_cls = (cls_score * objectness)[:, gt_labels]
+        p_pos = p_cls * p_loc
+        confidence_weight = torch.exp(p_pos * 3)
+        p_pos_weight = (confidence_weight * center_prior_weights) / (
+            (confidence_weight * center_prior_weights).sum(
+                0, keepdim=True)).clamp(min=EPS)
+        reweight_p_pos = (p_pos * p_pos_weight).sum(0)
+        pos_loss = F.binary_cross_entropy(
+            reweight_p_pos, torch.ones_like(reweight_p_pos), reduction='none')
+        pos_loss = pos_loss.sum() * self.pos_loss_weight
+        return pos_loss,
+
+    def get_neg_loss_single(self, cls_score, objectness, gt_labels, ious,
+                            inside_gt_bbox_mask):
+        """Calculate the negative loss of all points in feature map.
+
+        Args:
+            cls_score (Tensor): All category scores for each point on
+                the feature map. The shape is (num_points, num_class).
+            objectness (Tensor): Foreground probability of all points
+                and is shape of (num_points, 1).
+            gt_labels (Tensor): The zeros based label of all gt with shape of
+                (num_gt).
+            ious (Tensor): Float tensor with shape of (num_points, num_gt).
+              Each value represent the iou of pred_bbox and gt_bboxes.
+            inside_gt_bbox_mask (Tensor): Tensor of bool type,
+              with shape of (num_points, num_gt), each
+              value is used to mark whether this point falls
+              within a certain gt.
+
+        Returns:
+            tuple:
+            neg_loss (Tensor): The negative loss of all points
+                in the feature map.
+        """
+
+        joint_conf = (cls_score * objectness)
+        p_neg_weight = torch.ones_like(joint_conf)
+        temp_weight = (1 / (1 - ious))
+        iou_function_min = temp_weight.min(0)[0][None, :]
+        iou_function_max = temp_weight.max(0)[0][None, :]
+        neg_weight = 1 - (temp_weight - iou_function_min) / (
+            iou_function_max - iou_function_min).clamp(min=EPS)
+
+        foreground_idxs = torch.nonzero(inside_gt_bbox_mask, as_tuple=True)
+        neg_weight = neg_weight[inside_gt_bbox_mask]
+        p_neg_weight[foreground_idxs[0],
+                     gt_labels[foreground_idxs[1]]] = neg_weight
+
+        logits = (joint_conf * p_neg_weight)
+        neg_loss = (
+            logits**2 * F.binary_cross_entropy(
```

Comment:
Can we use focal loss interface?

Reply:
Because logits have been sigmoid calculation  if we want to use focal loss, we have to convert it to before sigmoid, there will
be a meaningless calculation.